src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -60,6 +58,32 @@ class ChannelInputStream extends InputStream {
         this.ch = ch;
     }
 
+    /*
+     * Reads bytes in Streams.DEFAULT_BUFFER_SIZE chunks until all requested
+     * bytes have been read or end-of-stream is reached.
+     * If the channel is selectable then it must be configured blocking.
+     */
+    private int readBytes(ByteBuffer bb) throws IOException {
+        int nread = 0;
+        int pos = bb.position();
+        int rem = bb.limit() - pos;
+        while (rem > 0) {
+            int size = Integer.min(rem, Streams.DEFAULT_BUFFER_SIZE);
+            bb.limit(pos + size);
+            int n = ch.read(bb);
+            if (n < 0) {
+                // EOF
+                if (nread == 0)
+                    nread = -1;
+                break;
+            }
+            nread += n;
+            pos += n;
+            rem -=n;
+        }
+        return nread;
```

Comment:
This still looks like a "read fully", I don't think we should have a loop here.

Reply:
So only `Streams.DEFAULT_BUFFER_SIZE` bytes are read per call?

Reply:
Yes.